src/git.ts

```diff
@@ -108,9 +108,25 @@ export async function getNewCommits(since?: Commit): Promise<CommitSequence> {
     if (since) {
         args.push(headBehindLastMigration ? 'HEAD..' + since.sha1 : since.sha1 + '..HEAD');
     }
-    const [stdout] = await execFile('git', args);
-    const output = stdout.toString().trim();
-    const commits = parseGitLog(output);
+    const commits = await (new Promise<CommitSequence>((resolve, reject) => {
+        const gitProcess = spawn('git', args);
+        let buffer = '';
+        let parsedCommits = new CommitSequence();
+        gitProcess.stdout.on('data', data => {
+            buffer += data.toString().trim();
+            const completeCommits = buffer.substring(0, buffer.lastIndexOf('>>>>COMMIT\n'));
+            buffer = buffer.substring(buffer.lastIndexOf('>>>>COMMIT\n'));
+            const parsedLog = parseGitLog(completeCommits);
+            parsedCommits = new CommitSequence(...parsedCommits, ...parsedLog);
+        });
+        gitProcess.on('error', reject);
+        gitProcess.on('exit', code => {
+            if (code !== 0) {
+                reject(new Error('git errored: ' + code));
+            }
+            resolve(new CommitSequence(...parsedCommits, ...parseGitLog(buffer)));
```

Comment:
I don't understand why there's code duplication here and in the data handler. Shouldn't this all have been added by the data handler by the time `exit` fires?

Reply:
We can't be sure a data event is the last, so there could be a commit left in the buffer which is only ensured to be complete if the process exits

Reply:
how did it get into the buffer if there was no data event emitted for it?

Reply:
We only parse commit messages we can be sure that are complete, so the next commit marker begins. For the last message, the only way to find out if the message is complete is to be sure the process exits, isn't it?

Reply:
depends on whether the commit marker is a delimiter or a terminator

Reply:
delimiter,

MARKER
msg
MARKER
msg
MARKER
msg